### PR TITLE
APP-972: Cross-sell notification support

### PR DESCRIPTION
**Pull request by hugokallstrom**
### Checklist

- [ ] Functionality is covered by an integration test
- [ ] Functionality is accessible in engineering mode

### CODE REVIEW

File: apollo/src/main/graphql/com/hedvig/android/owldroid/graphql/crossSells.graphql [hunk 1]
```diff
@@ -1,7 +1,41 @@
-query CrossSells {
+query CrossSells($locale: Locale!) {
   activeContractBundles {
     potentialCrossSells {
+      title
```
Comment (hedvigoscar) [inline apollo/src/main/graphql/com/hedvig/android/owldroid/graphql/crossSells.graphql:4]: Given the large amount of duplication we have now, should we use a GraphQL Fragment? Perhaps a separate query would be appropriate for the use-case? 🤔

File: app/src/main/java/com/hedvig/app/feature/crossselling/ui/CrossSellData.kt [hunk 2]
```diff
@@ -69,3 +104,10 @@ data class CrossSellData(
         )
     }
 }
+
+fun CrossSellType.toCrossSellType() = when (this) {
+    CrossSellType.ACCIDENT -> CrossSellData.CrossSellType.ACCIDENT
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/feature/crossselling/ui/CrossSellData.kt:109]: I think having us be explicit about the `CrossSellType` will erode some of the genericity of this solution - currently, any type of cross-sell is supported. If we enable `sealed class for enums` for the `CrossSellType`, we can simply match on the `rawValue`, and we never have to build in support client-side. 😄

File: app/src/main/java/com/hedvig/app/service/push/managers/CrossSellNotificationManager.kt [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+            val crossSell = getCrossSell(type)
+            if (crossSell != null) {
+                createIntentAndNotify(
+                    context = context,
+                    title = title,
+                    body = body,
+                    crossSell = crossSell
+                )
+            } else {
+                e { "Could not find cross-sell with type: $type" }
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/service/push/managers/CrossSellNotificationManager.kt:40]: Should we fall back to showing a notification, but simply opening to the insurance-tab? 🤔

File: app/src/main/java/com/hedvig/app/service/push/managers/CrossSellNotificationManager.kt [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    private fun parseCrossSellType(type: String?) = when (type) {
+        "ACCIDENT" -> CrossSellData.CrossSellType.ACCIDENT
+        "TRAVEL" -> CrossSellData.CrossSellType.TRAVEL
+        "HOME_CONTENT" -> CrossSellData.CrossSellType.HOME_CONTENT
+        else -> CrossSellData.CrossSellType.UNKNOWN
+    }
+
+    private val id = AtomicInteger(100)
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/service/push/managers/CrossSellNotificationManager.kt:102]: Do we need this `AtomicInteger`-thing for this type of notification? I was under the impression that it was only necessary for the chat-notifications 🤔

**Review by hedvigoscar — CHANGES_REQUESTED** [review]
Some architectural concerns to be considered before we go ahead, primarily the genericity-question 🙂

File: app/src/main/java/com/hedvig/app/feature/crossselling/ui/CrossSellData.kt (hunk 2 above)
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/feature/crossselling/ui/CrossSellData.kt:109]: Good idea

File: app/src/main/java/com/hedvig/app/service/push/managers/CrossSellNotificationManager.kt (hunk 3 above)
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/service/push/managers/CrossSellNotificationManager.kt:40]: Yes! 👍

File: apollo/src/main/graphql/com/hedvig/android/owldroid/graphql/crossSells.graphql (hunk 1 above)
Comment (hugokallstrom) [reply apollo/src/main/graphql/com/hedvig/android/owldroid/graphql/crossSells.graphql:4]: Yeah I suppose a fragment is the way to go. They are so clunky to work with though..

Comment (StylianosGakis) [reply apollo/src/main/graphql/com/hedvig/android/owldroid/graphql/crossSells.graphql:4]: What is it that you feel that Fragments are clunky?

Comment (hedvigoscar) [reply apollo/src/main/graphql/com/hedvig/android/owldroid/graphql/crossSells.graphql:4]: I agree that `object.fragments.someThingFragment.field` is less nice than `object.field` 😞 The tradeoff of not having to have duplicate methods acting on the data (which we do have a few times in this PR, mind you 😄) is worth it in my experience 🙂

Comment (hugokallstrom) [reply apollo/src/main/graphql/com/hedvig/android/owldroid/graphql/crossSells.graphql:4]: Its so unfortunate that you cant reference fragments without their "parent" query, like `CrossSells.Fragments.{...}` and `Insurance.Fragments.{...}` :( I agree its still worth it though. Fixing now!

**Review by StylianosGakis — APPROVED** [review]

File: app/src/main/java/com/hedvig/app/service/push/Util.kt [hunk 5]
```diff
@@ -25,5 +26,18 @@ fun setupNotificationChannel(
             }
         )
     }
+
+fun getImmutablePendingIntentFlags(): Int = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+    PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/service/push/Util.kt:31]: `FLAG_IMMUTABLE` can be specified on lower SDKs, right? 🤔

File: app/src/main/java/com/hedvig/app/service/push/Util.kt [hunk 6]
```diff
@@ -25,5 +26,18 @@ fun setupNotificationChannel(
...
+} else {
+    PendingIntent.FLAG_UPDATE_CURRENT
+}
+
+fun getMutablePendingIntentFlags(): Int = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
```
Comment (hedvigoscar) [inline app/src/main/java/com/hedvig/app/service/push/Util.kt:36]: 👍

**Review by hedvigoscar — APPROVED** [review]

**Pull request merged**